src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs

```diff
@@ -643,41 +639,29 @@ private ProcessingBranch PrepareProcessingBranch(Block suggestedBlock, Processin
                 break;
             }
 
-            if (isFastSyncTransition)
-            {
-                // If we hit this condition, it means that something is wrong in MultiSyncModeSelector.
-                // MultiSyncModeSelector switched to full sync when it shouldn't
-                // In this case, it is better to stop searching for more blocks and failed during the processing than trying to build a branch up to the genesis point
-                if (iterations > MaxBlocksDuringFastSyncTransition)
-                {
-                    if (_logger.IsWarn) _logger.Warn($"Too long branch to be processed during fast sync transition. Current block to be processed {toBeProcessed}, StateRoot: {toBeProcessed?.StateRoot}");
-                    break;
-                }
-
-                // if we have parent state it means that we don't need to go deeper
-                if (toBeProcessed?.StateRoot is null || _stateReader.HasStateForBlock(toBeProcessed.Header))
-                {
-                    if (_logger.IsInfo) _logger.Info($"Found state for parent: {toBeProcessed}, StateRoot: {toBeProcessed?.StateRoot}");
-                    break;
-                }
-                else
-                {
-                    if (_logger.IsDebug) _logger.Debug($"A new block {toBeProcessed} in fast sync transition branch - state not found");
-                }
-            }
-
             // TODO: there is no test for the second condition
             // generally if we finish fast sync at block, e.g. 8 and then have 6 blocks processed and close Neth
             // then on restart we would find 14 as the branch head (since 14 is on the main chain)
             // we need to dig deeper to go all the way to the false (reorg boundary) head
             // otherwise some nodes would be missing
             bool notFoundTheBranchingPointYet = !_blockTree.IsMainChain(branchingPoint.Hash!);
-            bool notReachedTheReorgBoundary = branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0);
+            bool notReachedTheReorgBoundary = branchingPoint.Number > (_blockTree.Head?.Header.Number);
+            bool hasState = toBeProcessed?.StateRoot is null || _stateReader.HasStateForBlock(toBeProcessed.Header);
```

Comment:
Can we make the logic here, so it would ALWAYS go deeper until it finds state or until it goes past our `Pruning.PersistenceInterval`? That recover so many issues. Would allow us to recover from long reorgs, would allow us to SetHead to random block on prunned nodes - node would figure out where it stored last state root.

Reply:
Done

Reply:
By keeping `notFoundTheBranchingPointYet` in line 635 we can break the loop even if we didn't find the state but processed the block earlier, which is exactly the case I wanted to handle. Let's ignore being on main chain if we don't have the state, until at least `PruningConfig.PersistanceInterval` is reached?